packages/xod-project/src/project.js

```diff
@@ -673,9 +673,16 @@ export const validatePatchContents = def(
         Patch.listLinks
       )(checkedPatch);
 
+    const checkIsPatchGeneric = R.compose(
+      R.any(Pin.isGenericPin),
+      Patch.listPins
+    );
+
     return checkNodeTypes(patch)
       .chain(checkedPatch => {
-        const deducedPinTypes = deducePinTypes(checkedPatch, project);
+        const deducedPinTypes = checkIsPatchGeneric(checkedPatch)
+          ? deducePinTypes(checkedPatch, project)
+          : {};
```

Comment:
I guess it would be better to incorporate the check *into* `deducePinTypes`. Otherwise, we’ll have code duplication or optimization misses.